README.md
Conversion of inactive links in the text
================

[<img src="https://skeeks.com/uploads/all/20/9e/a0/209ea08c3331e48573e6acbff8981990.png" alt="SkeekS blog" width="200"/>](https://skeeks.com/blog/programming/397-kak-preobrazovat-neaktivnye-ssylki-v-tekste-v-aktivnye-klikabelnye)

A small extension to convert active links in the text to active (clickable).
It can convert html or text, it is possible to convert on the client (js) as well as on the backend (php).

Небольшое расширение для преобразования неактивных ссылок в тексте в активные (кликабельные).
Может преобразовывать html или текст, возможно преобразовывать на клиенте (js) а так же на бэкенде (php).

[![Latest Stable Version](https://poser.pugx.org/skeeks/yii2-link-activation/v/stable.png)](https://packagist.org/packages/skeeks/yii2-link-activation)
[![Total Downloads](https://poser.pugx.org/skeeks/yii2-link-activation/downloads.png)](https://packagist.org/packages/skeeks/yii2-link-activation)

Installation
------------

```sh
$ composer require skeeks/yii2-link-activation "^0.0.3"
```

Or add this to your `composer.json` file:

```json
{
    "require": {
        "skeeks/yii2-link-activation": "^0.0.3"
    }
}
```

Client replace (on js)
-----

```php
<?
\skeeks\yii2\linkActivation\assets\TextHandlerAsset::register($this);
$this->registerJs(<<<JS
new sx.classes.LinkActivation(".description");
JS
);
?>
```
```html
<div class="description">
Какой то текст со ссылками https://test.ru,
https://google.ru/search
Все ссылки будут автоматически https://cms.skeeks.com/blog/releases/2-zapusk-sayta-dlya-skeeks-cms определены в этом тексте и станут кликабельными
</div>
```


Backend replace
-----

```php
<?
    $handler = new \skeeks\yii2\linkActivation\TextHandler();
    $handler->short_link_max_length = 45;
    echo $handler->replace($yourText);
?>
```

```php
<?= (new \skeeks\yii2\linkActivation\TextHandler())->replace($yourText); ?>
```

Screenshot
----------

[<img src="https://skeeks.com/uploads/all/7c/cb/dc/7ccbdc8a1393cdb88d87635b610fb108.png" alt="SkeekS blog" width="600"/>](https://skeeks.com/uploads/all/7c/cb/dc/7ccbdc8a1393cdb88d87635b610fb108.png)


Video
-----

[<img src="https://skeeks.com/uploads/all/e4/cf/10/e4cf1067b25a0ca815ea2c82b5703e71.png" alt="Video" width="557"/>](https://youtu.be/lzmuahaNwQQ)

Links
----------
* [Github](https://github.com/skeeks-semenov/yii2-link-activation)
* [Changelog](https://github.com/skeeks-semenov/yii2-link-activation/blob/master/CHANGELOG.md)
* [Issues](https://github.com/skeeks-semenov/yii2-link-activation/issues)
* [Packagist](https://packagist.org/packages/skeeks/yii2-link-activation)
* [SkeekS blog post](https://skeeks.com/blog/programming/397-kak-preobrazovat-neaktivnye-ssylki-v-tekste-v-aktivnye-klikabelnye)
* [SkeekS marketplace](https://cms.skeeks.com/marketplace/components/tools/other/396-preobrazovanie-neaktivnyh-ssylok-v-tekste)

___

> [![skeeks!](https://skeeks.com/img/logo/logo-no-title-80px.png)](https://skeeks.com)  
<i>SkeekS CMS (Yii2) — quickly, easily and effectively!</i>  
[skeeks.com](https://skeeks.com) | [cms.skeeks.com](https://cms.skeeks.com)



CHANGELOG.md
CHANGELOG
==============

0.0.3
-----------------
 * Can be used
 
0.0.2
-----------------
 * -
 
0.0.1
-----------------
 * -

src/components/TextHandler.php
<?php
/**
 * @link https://cms.skeeks.com/
 * @copyright Copyright (c) 2010 SkeekS
 * @license https://cms.skeeks.com/license/
 * @author Semenov Alexander <semenov@skeeks.com>
 */

namespace skeeks\yii2\linkActivation\components;

use yii\base\Component;

/**
 * @author Semenov Alexander <semenov@skeeks.com>
 */
class TextHandler extends Component
{

    /**
     * @var Maximum length of displayed short link
     */
    public $short_link_max_length = 45;

    /**
     * @param $text
     * @return null|string|string[]
     */
    public function replace($text)
    {
        $text = preg_replace("!<a.*?href=\"?'?http:\/\/([^ \"'>]+)\"?'?.*?>(.*?)</a>!is", "\\2", $text);
        $text = preg_replace("!<a.*?href=\"?'?https:\/\/([^ \"'>]+)\"?'?.*?>(.*?)</a>!is", "\\2", $text);
        /*return $text;*/
        //return preg_replace_callback('#(http|https|ftp|ftps|www)([://]|[\.])[^\s]+#ui', [$this, '_shortLink'], $text);
        return preg_replace_callback('/(((http|ftp|https):\/{2})+(([0-9a-z_-]+\.)+(aero|asia|biz|cat|com|coop|edu|gov|info|int|jobs|mil|mobi|museum|name|net|org|pro|tel|travel|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cd|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cx|cy|cz|cz|de|dj|dk|dm|do|dz|ec|ee|eg|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gg|gh|gi|gl|gm|gn|gp|gq|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|je|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mg|mh|mk|ml|mn|mn|mo|mp|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|nom|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ra|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sj|sk|sl|sm|sn|so|sr|st|su|sv|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|yu|za|zm|zw|arpa)(:[0-9]+)?((\/([~0-9a-zA-Z\#\+\%@\.\/_-]+))?(\?[0-9a-zA-Z\+\%@\/&\[\];=_-]+)?)?))\b/imuS', [$this, '_shortLink'], $text);
        //return preg_replace_callback("_(^|[\s.:;?\-\]<\(])(https?://[-\w;/?:@&=+$\|\_.!~*\|'()\[\]%#,☺]+[\w/#](\(\))?)(?=$|[\s',\|\(\).:;?\-\[\]>\)])_i", [$this, '_shortLink'], $text);
        //return preg_replace_callback("#\b(([\w-]+://?|www[.])[^\s()<>]+(?:\([\w\d]+\)|([^[:punct:]\s]|/)))#iS", [$this, '_shortLink'], $text);
        /*$reg = <<<TXT
_^(?:(?:https?|ftp)://)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,})))(?::\d{2,5})?(?:/[^\s]*)?$_iuS
TXT;

        return preg_replace_callback($reg, [$this, '_shortLink'], $text);*/
    }

    /**
     * @param $matches
     * @return string
     */
    private function _shortLink($matches)
    {
        $linkText = $matches[0];
        $linkData = parse_url($matches[0]);

        if (mb_strlen($linkText) > $this->short_link_max_length) {
            $linkText = mb_substr($linkText, 0, $this->short_link_max_length)."..";
        }
        if (preg_match('#(http|https|ftp|ftps)://[^\s]+#ui', $matches[0])) {

            $matches[0] = strip_tags($matches[0]);
            return '<a href="'.$matches[0].'" target="_blank" data-pjax="0" title="'.$matches[0].'">'.$linkText.'</a>';
        } else {

            $matches[0] = strip_tags($matches[0]);
            return '<a href="http://'.$matches[0].'" target="_blank" data-pjax="0" title="'.$matches[0].'">'.$linkText.'</a>';
        }
    }
}

src/componentns/TextToActiveLinks.php
<?php
/**
 * @link https://cms.skeeks.com/
 * @copyright Copyright (c) 2010 SkeekS
 * @license https://cms.skeeks.com/license/
 * @author Semenov Alexander <semenov@skeeks.com>
 */

namespace skeeks\yii2\textToActiveLinks\components;

use yii\base\Component;
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 */
class TextToActiveLinks extends Component
{

    /**
     * @var Maximum length of displayed short link
     */
    public $short_link_max_length = 45;

    /**
     * @param $text
     * @return null|string|string[]
     */
    public function correct($text)
    {
        $text = preg_replace("!<a.*?href=\"?'?http:\/\/([^ \"'>]+)\"?'?.*?>(.*?)</a>!is", "\\2", $text);
        $text = preg_replace("!<a.*?href=\"?'?https:\/\/([^ \"'>]+)\"?'?.*?>(.*?)</a>!is", "\\2", $text);
        /*return $text;*/
        //return preg_replace_callback('#(http|https|ftp|ftps|www)([://]|[\.])[^\s]+#ui', [$this, '_shortLink'], $text);
        return preg_replace_callback('/(((http|ftp|https):\/{2})+(([0-9a-z_-]+\.)+(aero|asia|biz|cat|com|coop|edu|gov|info|int|jobs|mil|mobi|museum|name|net|org|pro|tel|travel|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cd|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cx|cy|cz|cz|de|dj|dk|dm|do|dz|ec|ee|eg|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gg|gh|gi|gl|gm|gn|gp|gq|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|je|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mg|mh|mk|ml|mn|mn|mo|mp|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|nom|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ra|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sj|sk|sl|sm|sn|so|sr|st|su|sv|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|yu|za|zm|zw|arpa)(:[0-9]+)?((\/([~0-9a-zA-Z\#\+\%@\.\/_-]+))?(\?[0-9a-zA-Z\+\%@\/&\[\];=_-]+)?)?))\b/imuS', [$this, '_shortLink'], $text);
        //return preg_replace_callback("_(^|[\s.:;?\-\]<\(])(https?://[-\w;/?:@&=+$\|\_.!~*\|'()\[\]%#,☺]+[\w/#](\(\))?)(?=$|[\s',\|\(\).:;?\-\[\]>\)])_i", [$this, '_shortLink'], $text);
        //return preg_replace_callback("#\b(([\w-]+://?|www[.])[^\s()<>]+(?:\([\w\d]+\)|([^[:punct:]\s]|/)))#iS", [$this, '_shortLink'], $text);
        $reg = <<<TXT
_^(?:(?:https?|ftp)://)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,})))(?::\d{2,5})?(?:/[^\s]*)?$_iuS
TXT;

        return preg_replace_callback($reg, [$this, '_shortLink'], $text);
    }

    /**
     * @param $matches
     * @return string
     */
    private function _shortLink($matches)
    {
        $linkText = $matches[0];
        $linkData = parse_url($matches[0]);

        if (mb_strlen($linkText) > $this->short_link_max_length) {
            $linkText = mb_substr($linkText, 0, $this->short_link_max_length)."..";
        }
        if (preg_match('#(http|https|ftp|ftps)://[^\s]+#ui', $matches[0])) {

            $matches[0] = strip_tags($matches[0]);
            return '<a href="'.$matches[0].'" target="_blank" data-pjax="0" title="'.$matches[0].'">'.$linkText.'</a>';
        } else {

            $matches[0] = strip_tags($matches[0]);
            return '<a href="http://'.$matches[0].'" target="_blank" data-pjax="0" title="'.$matches[0].'">'.$linkText.'</a>';
        }
    }
}

src/assets/src/js/linkActivation.js
/**
 * @link https://cms.skeeks.com/
 * @copyright Copyright (c) 2010 SkeekS
 * @license https://cms.skeeks.com/license/
 * @author Semenov Alexander <semenov@skeeks.com>
 */

(function (sx, $, _) {
    /**
     *
     */
    sx.createNamespace('classes', sx);
    /**
     * Преобразование ссылок на клиенте
     */
    sx.classes.LinkActivation = sx.classes.Component.extend({

        /**
         * @param Selector
         * @param opts
         */
        construct: function (Selector, opts) {
            opts = opts || {};
            this.Selector = Selector || false;

            this.applyParentMethod(sx.classes.Component, 'construct', [opts]);
        },

        _onDomReady: function () {
            var self = this;

            if (this.Selector === false) {
                return false;
            }

            this.jQuerySelector = $(this.Selector);

            if ($.pjax) {
                $(document).on('pjax:complete', function (e) {
                    self._replace($(self.Selector, $(e.target)))
                    return true;
                });
            }

            this._replace(this.jQuerySelector);
        },

        _replace: function (jQuerySelector) {
            var self = this;

            jQuerySelector.each(function () {
                var jQueryTextWrapper = $(this);


                $("a[href^='http']", jQueryTextWrapper).wrap("<span class='sx-link'></span>");
                /*$("a[href^='https']", jQueryTextWrapper).wrapAll("<span class='sx-link'></span>");*/
                $("a[href^='ftp']", jQueryTextWrapper).wrap("<span class='sx-link'></span>");

                $(".sx-link", jQueryTextWrapper).each(function () {
                    var textLink = ("a", $(this)).text();
                    $(this).empty().append(textLink + " ");
                });

                var text = jQueryTextWrapper.html();
                //console.log(text);


                text = text.replace(new RegExp("<br>", 'g'), " <br> ");

                //var text = jQueryTextWrapper.html();
                var replaced = text.replace(self.getRegex(), '<a href="$&" target="_blank" data-pjax="0">$&</a>');
                jQueryTextWrapper.empty().append(replaced);
            });
        },

        /**
         * @returns {RegExp}
         */
        getRegex: function () {
            return new RegExp(
                // protocol identifier
                "(?:(?:https?|ftp)://)" +
                // user:pass authentication
                "(?:\S+(?::\S*)?@)?" +
                //"(?:\\S+(?::\\S*)?@)?" +
                "(?:" +
                // IP address exclusion
                // private & local networks
                "(?!(?:10|127)(?:\.\d{1,3}){3})" +
                "(?!(?:169\\.254|192\.168)(?:\.\d{1,3}){2})" +
                "(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})" +
                // IP address dotted notation octets
                // excludes loopback network 0.0.0.0
                // excludes reserved space >= 224.0.0.0
                // excludes network & broacast addresses
                // (first & last IP address of each class)
                "(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])" +
                "(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}" +
                "(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))" +
                "|" +
                // host name
                "(?:(?:[a-z0-9]-*)*[a-z0-9]+)" +
                //"(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)" +
                // domain name
                "(?:\.(?:[a-z0-9]-*)*[a-z0-9]+)*" + //~_
                //"(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*" + //Берутся пробелы в том числе


                // TLD identifier
                //"(?:\.(?:[a-z0-9]{2,}))" +
                //"(?:\.(?:[a-z0-9/?&_%=~-]{2,}))" + //http://sobirayka.ru.vps210.s7.h.skeeks.com/~sx/admin/admin-auth?_sx%5Bref%5D=%2F~sx

                //"(?:\.(?:[a-z0-9/?&;_%=~-]{2,}))" + //http://alekseevamv.ru.vps211.s7.h.skeeks.com/~sx/cms/admin-settings?component=&amp;component=v3project%5Cthemes%5Cmega%5CThemeMegaSettings
                //"(?:\.(?:[/]{2,}))" +
                // sorry, ignore TLD ending with dot
                //"\.?" +
                ")" +
                // port number
                "(?::\d{2,5})?"
                // resource path, excluding a trailing punctuation mark
                //+ "(?:[/?#](?:\S*[^\s!\"'()*,-.:;<>?\[\]_`{|}~]|))?"
                , "gi"
            );
        }

    });
})(sx, sx.$, sx._);